### ISSUE TITLE
Use Central Perspective Plane of Scanline search to do groundtoimage in LS

### ISSUE THREAD

**Issue by jessemapel** [USGS-Astrogeology/usgscsm#253]
Currently, we use the same secant search in image pixels that ISIS uses. We should instead use CPPS search which does a search in object space. [Here](https://pdfs.semanticscholar.org/e199/b33d5157e6134bd9fd387c0fb8816b8c1bf5.pdf) is the reference paper.

The rough outline of the algorithm is as follows:

At initialization of the sensor model compute the CPPs:
1. Compute the distorted focal plane position of each detector pixel.
1. Segment the curve into linear segments based on some tolerance (usually a function of the pixel size).
1. For each segment project it to a plane in object space at each scan-line

Also compute an affine approximate transformation from object space to image space.

When doing groundtoimage:
1. Compute the approximate image coordinate using the affine approximation.
1. Determine which segment of the detector to search.
1. Compute the distance from the first CPP to the ground point.
1. Use the average distance between CPPs to get an approximate preceding CPP.
1. Do a secant search on the distance from the ground to the preceding CPP to find the actual preceding CPP.
1. Interpolate the sub-scanline.
1. Compute the back projected distorted focal plane coordinate of the ground point using the scaline's exterior orientation.
1. Do a correction to account for the linear approximation of the detector in the distorted focal plane.
1. Compute the final image space coordinates of the ground point using the final scan line.

In short, pre-compute look vectors and then search in those instead of searching in image space.

**Comment by jessemapel** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Determine how to implement CPPS search algorithm](https://trello.com/c/zVxQLaBY/8-determine-how-to-implement-cpps-search-algorithm)

**Comment by jessemapel** [comment]
Unfortunately, CPPS is not going to be practical for CSM because of the adjustable parameters. Passing parameter adjustments requires the planes be recomputed. For position and orientation changes this is not an issue, but for focal length changes or distortion changes, this could be quite severe and would result in significant performance loss.
